File: Code/glass.js
/*
...
*/

// size
Width = 600;
Height = 500;
Area = Width*Height;

// --- pre-specified order
/**
order = [[300,100],[250,125],[150,150],[150,150],[130,90],[130,90]]; // [width,height]
orderArea = totalArea(order);
**/

// --- glaslijst

var lines = document.getElementById("glaslijst").innerHTML.split("\n");
order = [];
for (var i = 0; i < lines.length; i++){
	var sizes = lines[i].split("\t");
	order[i] = [parseInt(sizes[0]),parseInt(sizes[1])]
}
console.log(order);
orderArea = totalArea(order);


// --- glaslijst (tsv)
/**
var order = [];
d3.tsv("https://raw.githubusercontent.com/Pvtwuyver/programmeertheorie/master/GlasLijsten/GlasLijst1.tsv", function(error, data) {
	if (error) throw error;
	// Coerce the data to numbers.
	data.forEach(function(d) {
	d.w = +d.w;
	d.h = +d.h;
	order.push([d.w,d.h]);
	});
	console.log(order);
	orderArea = totalArea(order);
})
**/

// --- random order
/**
orderArea = 0;
count = 0;
order = [];

while (orderArea <= Area) {
	var randomWidth = Math.round(randomSize()), randomHeight = Math.round(randomSize());
	randomCut = [randomWidth,randomHeight]
	orderArea += randomWidth*randomHeight;
	order[count] = randomCut;
	count++;
}
**/

total = order.length;
console.log("Number of cuts ordered: ",total, orderArea, Area);

// meerdere platen

numberOfWindows = 0;
while (order.length > 0) {

var svg = d3.select("body").append("div")
	.attr("style","width: "+ Width + "px; height: "+ Height + "px; border: 2px solid black;  background-color: #eee; transform: translate(0px,"+ numberOfWindows*10 +"px);")
   .append("svg")
   	.attr("height",Height)
	.attr("width",Width);

verticalFill(Width,Height,0,0);
//horizontalFill(Width,Height,0,0);

usedArea = orderArea - totalArea(order);
efficiency = (usedArea/Area);
console.log("Efficiency: ",efficiency);
	
numberOfWindows++;
}


// één plaat
/**
var svg = d3.select("body").append("div")
	.attr("style","width: "+ Width + "px; height: "+ Height + "px; border: 2px solid black;  background-color: #eee;")
   .append("svg")
   	.attr("height",Height)
	.attr("width",Width);
verticalFill(Width,Height,0,0);
**/

usedArea = orderArea - totalArea(order);
efficiency = (usedArea/Area);
console.log("Upper bound: ",orderArea/Area);
console.log("Number of cuts made: ",total - order.length, "/", total);
console.log("Efficiency: ",efficiency);
console.log(order);

function verticalFill(W, H, xIndex, yIndex) {
	var xStart = xIndex;
	var yStart = yIndex;
	var colWidth = W, rowHeight = H;
	while (order.length > 0) {
		orientation(order, "width");
		console.log("vertical - colWidth:", colWidth, "rowHeight", rowHeight);
		cut = search(order, colWidth, rowHeight);
		console.log("Cut: ", cut);
		if (cut != null) { var cutWidth = cut[0], cutHeight = cut[1];}
		else {
			console.log("new col", rowHeight);
			orientation(order, "height");
			cut = search(order, colWidth, rowHeight);
			if (cut != null) { var cutWidth = cut[0], cutHeight = cut[1];}
			else {
				orientation(order, "width");
				xIndex += colWidth;
				console.log("x: ", xIndex);
				colWidth = W - (xIndex - xStart);
				if (colWidth < 0) { console.log("break / no room");break;}
				yIndex = yStart;
				rowHeight = H;
				/**
				if ((rowHeight > colWidth) && (colWidth > 0) && (rowHeight > 0)) {
					horizontalFill(colWidth,rowHeight,xIndex,yIndex);
					orientation(order, "width");
				}
				**/
				console.log("next cut: ",colWidth, rowHeight);
				cut = search(order, colWidth, rowHeight);
				console.log("Cut: ", cut);
				if (cut != null) { var cutWidth = cut[0], cutHeight = cut[1];}
				else {
					console.log("break/ no fit");
					orientation(order, "height");
					console.log("next cut: ",colWidth, rowHeight);
					cut = search(order, colWidth, rowHeight);
					if (cut != null) { var cutWidth = cut[0], cutHeight = cut[1];}
					else {break;};
				//---
				}	
			}
		};
		svg.append("rect")
			.attr("transform", "translate(" + xIndex + "," + yIndex + ")")
			.attr("width",cutWidth)
			.attr("height",cutHeight)
			.attr("style","fill: #efda62; stroke-width:1; stroke:black");
		if (yIndex == yStart) {
			colWidth = cutWidth;
		}
		console.log("Room left:", (colWidth - cutWidth), cutHeight);
		if ((cutHeight > (colWidth - cutWidth)) && ((colWidth - cutWidth) > 0)) {
			horizontalFill(colWidth - cutWidth, cutHeight, xIndex + cutWidth, yIndex);
			orientation(order, "width");
		}
		else if ((cutHeight <= (colWidth - cutWidth)) && ((colWidth - cutWidth) > 0)) {
			verticalFill(colWidth - cutWidth, cutHeight, xIndex + cutWidth, yIndex);
		}
		else {console.log("No recursion possible");};
		yIndex += cutHeight;
		rowHeight = H - (yIndex - yStart);
		//if (rowHeight > colWidth) {horizontalFill(colWidth, rowHeight, xIndex, yIndex)};
	}
};

function horizontalFill(W, H, xIndex, yIndex) {
	var xStart = xIndex;
	var yStart = yIndex;
	var colWidth = W, rowHeight = H;
	while (order.length > 0) {
		orientation(order, "height");
		console.log("horizontal - colWidth:", colWidth, "rowHeight", rowHeight);
		cut = search(order, colWidth, rowHeight);
		console.log("Cut: ", cut);
		if (cut != null) { var cutWidth = cut[0], cutHeight = cut[1];}
		else {
			console.log("new row");
			orientation(order, "height");
			cut = search(order, colWidth, rowHeight);
			if (cut != null) { var cutWidth = cut[0], cutHeight = cut[1];}
			else {
				yIndex += rowHeight;
				rowHeight = H - (yIndex - yStart);
				if (rowHeight < 0) { console.log("break / no room");break;}
				xIndex = xStart;
				colWidth = W;
				/**
				if ((colWidth > rowHeight) && (colWidth > 0) && (rowHeight > 0)) {
					verticalFill(colWidth,rowHeight,xIndex,yIndex);
					orientation(order, "height");
				}
				**/
				cut = search(order, colWidth, rowHeight);
				console.log("Cut: ", cut);
				if (cut != null) { var cutWidth = cut[0], cutHeight = cut[1];}
				else {
					console.log("break/ no fit");
					orientation(order, "width");
					cut = search(order, colWidth, rowHeight);
					if (cut != null) { var cutWidth = cut[0], cutHeight = cut[1];}
					else {break;};
				}
			}
		};
		svg.append("rect")
			.attr("transform", "translate(" + xIndex + "," + yIndex + ")")
			.attr("width",cutWidth)
			.attr("height",cutHeight)
			.attr("style","fill: #efda62; stroke-width:1; stroke:black");
		if (xIndex == xStart) {
			rowHeight = cutHeight;
		}
		console.log("Room left:", cutWidth, (rowHeight - cutHeight));
		if ((cutWidth > (rowHeight - cutHeight)) && ((rowHeight - cutHeight) > 0)) {
			verticalFill(cutWidth, rowHeight - cutHeight, xIndex, yIndex + cutHeight);
			orientation(order, "height");
		}
		else if ((cutWidth <= (rowHeight - cutHeight)) && ((rowHeight - cutHeight) > 0)) {
			horizontalFill(cutWidth, rowHeight - cutHeight, xIndex, yIndex + cutHeight);
		}
		else {console.log("No recursion possible");};
		xIndex += cutWidth;
		colWidth = W - (xIndex - xStart);
	}
};

// sort list by width
function sortWidth(a, b) {
    if (a[0] === b[0]) {
        return 0;
    }
    else {
        return (a[0] > b[0]) ? -1 : 1;
    }
}

// sort list by height
function sortHeight(a, b) {
    if (a[1] === b[1]) {
        return 0;
    }
    else {
        return (a[1] > b[1]) ? -1 : 1;
    }
}

// sort list by circumference
function sortCirc(a, b) {
	circA = 2*a[0] + 2*a[1];
	circB = 2*b[0] + 2*b[1];
    if (circA === circB) {
        return 0;
    }
    else {
        return (circA > circB) ? -1 : 1;
    }
}

// change the orientation of a glass cut
function orientation(list, side) {
	console.log("change orientation", side);
	for (item in list) {	
		var max = Math.max(list[item][0],list[item][1]), min = Math.min(list[item][0],list[item][1]);
		if (side == "width") {
			list[item][0] = max; list[item][1] = min;
			list.sort(sortHeight);
			list.sort(sortWidth);
		}
		else if (side == "height") {
			list[item][1] = max; list[item][0] = min;
			list.sort(sortWidth);
			list.sort(sortHeight);
		}
	}
}

// search
function search(list, maxWidth, maxHeight) {
	for (var i in list) {
		if ((list[i][0] <= maxWidth) && (list[i][1] <= maxHeight)) {
			dim = list[i];
			list.splice(i,1);
			return dim
		}
	}
	return null;
}

function totalArea(list) {
	var area = 0;
	for (i in list) {
		area += list[i][0]*list[i][1];
	}
	return area;
}

function randomSize() {
	var value = 0;
	while (value < 20) {
		value = 350*((Math.random() + Math.random() + Math.random() + Math.random() + Math.random() + Math.random()) - 3) / 3;
	}	
	return Math.abs(value);
}

File: PythonCode/glass.py
from math import *
import matplotlib.pyplot as plt
import matplotlib.patches as patches

def totalArea(list):
	area = 0
	for i in list:
		area += i[0]*i[1]
	return area

def sortWidth(list):
	sortedList = sorted(list, key=lambda x: (x[0],x[1]), reverse=True)
	return sortedList
	
def sortHeight(list):
	sortedList = sorted(list, key=lambda x: (x[1],x[0]), reverse=True)
	return sortedList
	
def orientation(list, side):
	if side == "width":
		for item in list:
			bigSide, smallSide = max(item[0],item[1]), min(item[0],item[1])
			item[0], item[1] = bigSide, smallSide
		list = sortWidth(list)
		return list
	elif side == "height":
		for item in list:
			bigSide, smallSide = max(item[0],item[1]), min(item[0],item[1])
			item[0], item[1] = smallSide, bigSide
		list = sortHeight(list)
		return list

def search(list, maxWidth, maxHeight):
	for i in list:
		if (i[0] <= maxWidth) and (i[1] <= maxHeight):
			list.remove(i)
			return i
	return None 
		
def verticalFill(list, W, H, xIndex, yIndex):
	xStart = xIndex
	yStart = yIndex	
	colWidth, rowHeight = W, H
	while len(list) > 0:
		# initial search
		list = orientation(list, "width")
		print("search: ", colWidth, rowHeight, " In list: ", list)
		cut = search(list, colWidth, rowHeight)
		if cut is not None: 
			cutWidth, cutHeight = cut[0], cut[1]
		else:
			# try different orientation
			list = orientation(list, "height")
			cut = search(list, colWidth, rowHeight)
			if cut is not None:
				cutWidth, cutHeight = cut[0], cut[1]
			else:
				# new column needed
				print("new col")
				list = orientation(list, "width")
				xIndex += colWidth
				colWidth = W - (xIndex - xStart)
				yIndex = yStart
				rowHeight = H
				#print("search: ", colWidth, rowHeight)
				cut = search(list, colWidth, rowHeight)
				if cut is not None:
					cutWidth, cutHeight = cut[0], cut[1]
				else:
					list = orientation(list, "height")
					cut = search(list, colWidth, rowHeight)
					if cut is not None: 
						cutWidth, cutHeight = cut[0], cut[1]
					else:
						break
		print("cut: ", cut)
		ax1 = fig1.add_subplot(111, aspect='equal')
		ax1.add_patch(
			patches.Rectangle(
				(xIndex, yIndex),   # (x,y)
				cutWidth,          # width
				cutHeight,          # height
				facecolor = "#efda62"
			)
		)
		if yIndex == yStart:
			colWidth = cutWidth
		# room left?
		
		if (cutHeight > (colWidth - cutWidth)) and ((colWidth - cutWidth) > 0):
			print("list for recursion: ", list)
			list = horizontalFill(list, colWidth - cutWidth, cutHeight, xIndex + cutWidth, yIndex)
			list = orientation(list, "width")
		elif (cutHeight <= (colWidth - cutWidth)) and ((colWidth - cutWidth) > 0):
			print("list for recursion: ", list)
			list = verticalFill(list, colWidth - cutWidth, cutHeight, xIndex + cutWidth, yIndex)
		
		yIndex += cutHeight
		rowHeight = H - (yIndex - yStart)
	return list
	
def horizontalFill(list, W, H, xIndex, yIndex):
	xStart = xIndex
	yStart = yIndex	
	colWidth, rowHeight = W, H
	while len(list) > 0:
		print(list)
		# initial search
		list = orientation(list, "height")
		print("search: ", colWidth, rowHeight)
		cut = search(list, colWidth, rowHeight)
		if cut is not None: 
			cutWidth, cutHeight = cut[0], cut[1]
		else:
			# try different orientation
			list = orientation(list, "width")
			cut = search(list, colWidth, rowHeight)
			if cut is not None:
				cutWidth, cutHeight = cut[0], cut[1]
			else:
				# new row needed
				print("new row")
				list = orientation(list, "height")
				yIndex += rowHeight
				rowHeight = H - (yIndex - yStart)
				xIndex = xStart
				colWidth = W
				#print("search: ", colWidth, rowHeight)
				cut = search(list, colWidth, rowHeight)
				if cut is not None:
					cutWidth, cutHeight = cut[0], cut[1]
				else:
					list = orientation(list, "width")
					cut = search(list, colWidth, rowHeight)
					if cut is not None: 
						cutWidth, cutHeight = cut[0], cut[1]
					else:
						break
		print("cut: ", cut)
		ax1 = fig1.add_subplot(111, aspect='equal')
		ax1.add_patch(
			patches.Rectangle(
				(xIndex, yIndex),   # (x,y)
				cutWidth,          # width
				cutHeight,          # height
				facecolor = "#efda62"
			)
		)
		if xIndex == xStart:
			rowHeight = cutHeight
		# room left?
		if (cutWidth > (rowHeight - cutHeight)) and ((rowHeight - cutHeight) > 0):
			list = verticalFill(list, cutWidth, rowHeight - cutHeight, xIndex, yIndex + cutHeight)
			list = orientation(list, "height")
		elif (cutWidth <= (rowHeight - cutHeight)) and ((rowHeight - cutHeight) > 0):
			list = horizontalFill(list, cutWidth, rowHeight - cutHeight, xIndex, yIndex + cutHeight)

		xIndex += cutWidth
		colWidth = W - (xIndex - xStart)
	return list

Width = 600
Height = 500
Area = Width*Height	

# glaslijst 1
#order = [[190,270],[90,160],[120,290],[110,220],[160,120],[90,120],[200,100],[110,290],[120,170],[100,320],[90,160],[190,300],[170,250],[180,340],[170,180],[90,100],[110,270],[70,220],[40,130],[140,330],[130,110],[40,240]]
# glaslijst 2
#order = [[150,150],[50,110],[160,270],[130,270],[130,130],[190,160],[200,190],[170,240],[110,220],[110,140],[70,230],[140,170],[160,240],[200,130],[150,100],[190,220],[60,150],[40,240]]
# glaslijst 3:
#order = [[110,100],[130,240],[130,220],[90,160],[40,100],[50,140],[150,250],[70,200],[160,120],[120,120],[100,190],[190,240],[120,270],[60,130],[160,230],[170,170],[200,170],[90,210],[60,190],[120,180],[110,190],[180,270],[160,120],[160,100]]
# glaslijst 4:
#order = [[90,220],[110,260],[80,120],[80,280],[50,280],[80,270],[160,190],[40,190],[90,250],[180,210],[180,250],[170,270],[140,230],[110,270],[80,140],[100,270],[140,210],[120,200],[120,150]]
# glaslijst 2,3 en 4
# order = [[150,150],[50,110],[160,270],[130,270],[130,130],[190,160],[200,190],[170,240],[110,220],[110,140],[70,230],[140,170],[160,240],[200,130],[150,100],[190,220],[60,150],[40,240],[110,100],[130,240],[130,220],[90,160],[40,100],[50,140],[150,250],[70,200],[160,120],[120,120],[100,190],[190,240],[120,270],[60,130],[160,230],[170,170],[200,170],[90,210],[60,190],[120,180],[110,190],[180,270],[160,120],[160,100],[90,220],[110,260],[80,120],[80,280],[50,280],[80,270],[160,190],[40,190],[90,250],[180,210],[180,250],[170,270],[140,230],[110,270],[80,140],[100,270],[140,210],[120,200],[120,150]]
# type 3:
order = [[70,180],[180,100],[90,130],[70,180],[150,120],[60,200],[80,200],[60,240],[110,160],[120,240],[110,170],[150,230]]

beginArea = totalArea(order)
wasteArray = []
numberOfSheets = 0

while len(order) > 0:
	fig1 = plt.figure()
	order = verticalFill(order, Width, Height, 0, 0)
	waste = Area - (beginArea - totalArea(order))
	beginArea = beginArea - (beginArea - totalArea(order))
	wasteArray.append(waste)
	plt.axis([0,Width,0,Height])
	plt.tick_params(axis='both', which='both', bottom='off', top='off', labelbottom='off', right='off', left='off', labelleft='off')
	plt.plot()
	numberOfSheets += 1
plt.show()

score = (Area - (sum(wasteArray[0:-1])/numberOfSheets))/float(Area)

print("Score: ", score)